### PR TITLE
MF-201 - Replace ORM with SQL

### LINKED ISSUE #201: Use raw SQL to interact with PostgreSQL
Instead of relying on GORM, all queries and database migrations should be specified using plain SQL. References:
- [pq](https://github.com/lib/pq)
- [migrate](https://github.com/mattes/migrate)

### PR DESCRIPTION
**Pull request by mijicd**
This pull request resolves #201.

Besides replacing GORM completely, this pull request introduces in-memory specified SQL database migrations via [sql-migrate](https://github.com/rubenv/sql-migrate).

### CODE REVIEW

File: clients/postgres/channels.go [hunk 1]
```diff
@@ -1,120 +1,183 @@
...
+		return err
+	}
 
-	if res.Error == nil && res.RowsAffected == 0 {
+	if cnt == 0 {
 		return clients.ErrNotFound
 	}
 
-	return res.Error
+	return err
```
Comment (anovakovic01) [inline clients/postgres/channels.go]: Return `nil` here.

File: clients/postgres/channels.go [hunk 2]
```diff
@@ -1,120 +1,183 @@
...
 }
 
 func (cr channelRepository) One(owner, id string) (clients.Channel, error) {
-	channel := clients.Channel{}
+	q := `SELECT name FROM channels WHERE id = $1 AND owner = $2`
+	channel := clients.Channel{ID: id, Owner: owner}
+	err := cr.db.QueryRow(q, id, owner).Scan(&channel.Name)
+
+	if err != nil {
```
Comment (anovakovic01) [inline clients/postgres/channels.go]: `err` declaration can be part of `if`.

File: clients/postgres/channels.go [hunk 3]
```diff
@@ -1,120 +1,183 @@
...
+		cr.log.Error(fmt.Sprintf("Failed to retrieve connected due to %s", err))
+		return clients.Channel{}, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		c := clients.Client{Owner: owner}
+		err = rows.Scan(&c.ID, &c.Name, &c.Type, &c.Key, &c.Payload)
+		if err != nil {
```
Comment (anovakovic01) [inline clients/postgres/channels.go]: `err` declaration can be part of `if`.

File: clients/postgres/channels.go [hunk 4]
```diff
@@ -1,120 +1,183 @@
...
+	if err != nil {
+		cr.log.Error(fmt.Sprintf("Failed to retrieve channels due to %s", err))
+		return []clients.Channel{}
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		c := clients.Channel{Owner: owner}
+		err = rows.Scan(&c.ID, &c.Name)
+		if err != nil {
```
Comment (anovakovic01) [inline clients/postgres/channels.go]: `err` declaration can be part of `if`.

File: clients/postgres/channels.go [hunk 5]
```diff
@@ -1,120 +1,183 @@
...
-	EXISTS (SELECT 1 FROM clients WHERE owner = ? AND id = ?);`
-
-	res := cr.db.Exec(sql, chanID, clientID, owner, chanID, owner, clientID)
-
-	if res.Error == nil && res.RowsAffected == 0 {
-		return clients.ErrNotFound
+	q := `INSERT INTO connections (channel_id, channel_owner, client_id, client_owner) VALUES ($1, $2, $3, $2)`
+
+	_, err := cr.db.Exec(q, chanID, owner, clientID)
+	if err != nil {
```
Comment (anovakovic01) [inline clients/postgres/channels.go]: `err` declaration can be part of `if`.

File: clients/postgres/channels.go [hunk 6]
```diff
@@ -1,120 +1,183 @@
...
-	var exists bool
-	if err := row.Scan(&exists); err != nil {
-		// TODO: this error should be logged
-		return false
-	}
+	q := "SELECT EXISTS (SELECT 1 FROM connections WHERE channel_id = $1 AND client_id = $2);"
 
+	// TODO: log errors?
+	exists := false
+	cr.db.QueryRow(q, chanID, clientID).Scan(&exists)
```
Comment (anovakovic01) [inline clients/postgres/channels.go]: I think that this error should be logged. There is not other way for us to know that something went wrong.

File: clients/postgres/clients.go [hunk 7]
```diff
@@ -1,68 +1,102 @@
...
+	if err != nil {
+		cr.log.Error(fmt.Sprintf("Failed to retrieve clients due to %s", err))
+		return []clients.Client{}
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		c := clients.Client{Owner: owner}
+		err = rows.Scan(&c.ID, &c.Name, &c.Type, &c.Key, &c.Payload)
+		if err != nil {
```
Comment (anovakovic01) [inline clients/postgres/clients.go]: `err` declaration can be part of `if`.

File: clients/postgres/init.go [hunk 8]
```diff
@@ -1,40 +1,72 @@
...
+						type	VARCHAR(10)  NOT NULL,
+						name	TEXT,
+						key		TEXT,
+						payload TEXT,
+						PRIMARY KEY (id, owner)
+					)`,
+					`CREATE TABLE channels (
+						id	  CHAR(36),
+						owner VARCHAR(254),
+						name  TEXT,
```
Comment (anovakovic01) [inline clients/postgres/init.go]: Align field types.

File: coap/adapter.go [hunk 9]
```diff
@@ -99,7 +126,7 @@ func (svc *adapterService) Unsubscribe(clientID string) {
 func (svc *adapterService) SetTimeout(clientID string, timer *time.Timer, duration int) (chan bool, error) {
 	sub, ok := svc.get(clientID)
 	if !ok {
-		return nil, errEntryNotFound
+		return nil, errors.New("observer entry not founds")
```
Comment (anovakovic01) [inline coap/adapter.go]: Fix typo. Should this error be public?

File: users/service.go [hunk 10]
```diff
@@ -36,3 +36,47 @@ type Service interface {
...
+
+type usersService struct {
+	users  UserRepository
+	hasher Hasher
+	idp    IdentityProvider
+}
+
+// New instantiates the users service implementation.
+func New(users UserRepository, hasher Hasher, idp IdentityProvider) Service {
+	return &usersService{users, hasher, idp}
```
Comment (anovakovic01) [inline users/service.go]: Use named fields.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

File: coap/adapter.go (hunk 9 above)
Comment (mijicd) [reply coap/adapter.go]: No need for that. It wasn't exported before, and this its only usage.

**Review by anovakovic01 — APPROVED** [review]
LGTM!

**Pull request merged**